### DISCUSSION TITLE
Function arguments vs action data filter

### DISCUSSION THREAD

**Discussion by mswiderski** [serverlessworkflow/specification#558] [General]
Hi,

i am looking into functions within action state and started to think what kind of relationship there is between function’s arguments and action data filters?
Arguments are not really described in the spec in that manner so not sure how others interpret it, more specificly:
- is it really needed when we have already state data and the action data that be be setup as needed nt action data filter
- Can arguments access any level of data or only enclosing action after filter has been applied (if was defined otherwise access to state data)

I started to think that arguments might be unneeded level of data and functions could simply get action data based on the filter. 

What do you guys think? Or how other runtimes approach this part of the spec?

**Comment by tsurdilo** [comment]
> is it really needed when we have already state data and the action data that be be setup as needed nt action data filter

Imo yes, besides filtering data that action params can access, action data filter also provides the useResults and toStateData
properties that are useful too.
One thing I wanted to add to spec is reusable action filters, which would be imo super useful for testing scenarios (apply a different filter on state data to test out different values for props for example)

> Can arguments access any level of data or only enclosing action after filter has been applied (if was defined otherwise access to state data)

If action filter fromStateData is specified, then its results define the json that can be used in properties. If not specified, yes action props can use the whole state data

**Comment by mswiderski** [comment]
I agree to all settings around action data filter, what I am still not clear about is the arguments of the function that are set on 
https://github.com/serverlessworkflow/specification/blob/main/specification.md#FunctionRef-Definition

Is it only needed for graphql or all? What got me thinking is that arguments in the above link are again expressions so it seems to be yet another filter expression. My main question is action data vs arguments - what’s the difference and why we need both?

**Reply by mswiderski** [reply]
@tsurdilo any comments on the above?

**Comment by ricardozanini** [comment]
@mswiderski I'm closing this one since it's outdated due to the new version we have now if you don't mind taking a look: https://github.com/serverlessworkflow/specification/blob/main/dsl.md
